## Analysis of High Dimensional Data - Lab 2
## Unsupervised learning
### SVD (singular value decomposition) 


###### Dr. Chamberlain Mbah --yields.io

**The purpose of the following exercises is mainly to get more familiar with SVD, and with
working with vectors and matrices in python. Perform all the tasks in python (i.e., Mathematical proofs
are not needed)**

#### First we have to import relevant libraries

In [74]:
import os #for operating system information such as working directory
import numpy as np
import pandas as pd # library in python which help us to read the data
import matplotlib.pyplot as plt # for  ploting graph
# We will import linear regression and two penalised regressions lasso and ridge
from sklearn.linear_model import LinearRegression, Lasso, Ridge

In [75]:
#Set your working dirrection to the location of your data
os.getcwd() # or pwd

'/home/samuel/AIMS_CAMEROON_2022_2023/HIGH_DIMENSION_DATA _ANALYSIS/Assignment_High_Dimensional_Data_Analysis'

In [76]:
#Data path 
pathData='/home/samuel/AIMS_CAMEROON_2022_2023/HIGH_DIMENSION_DATA _ANALYSIS/Assignment_High_Dimensional_Data_Analysis'

# Read in the data 


In [77]:
# read data into a DataFrame
X=pd.read_csv(pathData+'/cheese.csv',index_col=0)

## Exercises 

1.) Let's view the first 10 observations in the dataset 

In [78]:
X.head(10) # note that by default i.e.,python print just the five first observations

taste  Acetic    H2S  Lactic
Case                              
1      12.3   4.543  3.135    0.86
2      20.9   5.159  5.043    1.53
3      39.0   5.366  5.438    1.57
4      47.9   5.759  7.496    1.81
5       5.6   4.663  3.807    0.99
6      25.9   5.697  7.601    1.09
7      37.3   5.892  8.726    1.29
8      21.9   6.078  7.966    1.78
9      18.1   4.898  3.850    1.29
10     21.0   5.242  4.174    1.58

2.) The command to view the tail 10 observations in the dataset is the following:

In [79]:
X.tail(10) # this command help us to view the 10 last rows of our dataset

taste  Acetic     H2S  Lactic
Case                               
21     14.0   4.564   4.949    1.15
22     15.2   5.298   5.220    1.33
23     32.0   5.455   9.242    1.44
24     56.7   5.855  10.199    2.01
25     16.8   5.366   3.664    1.31
26     11.6   6.043   3.219    1.46
27     26.5   6.458   6.962    1.72
28      0.7   5.328   3.912    1.25
29     13.4   5.802   6.685    1.08
30      5.5   6.176   4.787    1.25

3.)Command to get the dimension of any data:  

In [80]:
X.shape

(30, 4)

This is not a high-dimensional data because the number of observations (n=30) is greater than the number of feartures(p=5)

4.)Command to view a summary of a data by using the transpose:

In [81]:
X.describe().T #  summary of  data, T means transpose

count       mean        std    min      25%     50%       75%     max
taste    30.0  24.533333  16.255383  0.700  13.5500  20.950  36.70000  57.200
Acetic   30.0   5.498033   0.570878  4.477   5.2375   5.425   5.88275   6.458
H2S      30.0   5.941767   2.126879  2.996   3.9775   5.329   7.57475  10.199
Lactic   30.0   1.442000   0.303490  0.860   1.2500   1.450   1.66750   2.010

5.) Scale your data, i.e., substract the mean from each column and 
divide by its standard deviation

In [82]:
mean=X.mean()
sds=X.std()
X_scaled=(X-mean)/sds

6.) Verify that X is indeed scaled

In [83]:
X_scaled.describe().T 

count          mean  std       min       25%       50%       75%  \
taste    30.0  2.960595e-16  1.0 -1.466181 -0.675674 -0.220440  0.748470   
Acetic   30.0 -2.176037e-15  1.0 -1.788530 -0.456373 -0.127932  0.673903   
H2S      30.0  2.257453e-16  1.0 -1.385018 -0.923544 -0.288106  0.767784   
Lactic   30.0  8.770762e-16  1.0 -1.917691 -0.632640  0.026360  0.743023   

             max  
taste   2.009591  
Acetic  1.681561  
H2S     2.001634  
Lactic  1.871561

As the mean tend to zero and the standart deviation equal to one we can conclude that our data is centered and normalize.

6a.) Rank of X:

In [84]:
np.linalg.matrix_rank(X_scaled) # which linalg is linear algebra and matrix_rank give us the rank of the matrix
# which is X_scaled

4

7.) Let's perform SVD on X as we saw in the course notes and
store our results as $U$, $D$, and $V\_tranpose$

In [85]:
U,D,V_transpose=np.linalg.svd(X_scaled,full_matrices=False) #if we didn't put full_matices, it means that our matix 
# is in his complete or total form and the result we are going to get is by default. Si recommend to put 
# full_matrices=False, that option implement an optimal reduced rank approximation of a matrix
U.shape, D.shape,V_transpose.shape # to get the dimension of each matrix and we can observe that the forms are 
#consistent,we can do multiplication on them

((30, 4), (4,), (4, 4))

8.) Let us show that $\mathbf u_1$ is a normalised vector;the same for $\mathbf u_2$.
Let's show too that $\mathbf u_1$ and $\mathbf u_2$  are orthogonal vectors. Then let's show the 
orthonormality of all vectors $\mathbf u_j$ in a single calculation (by using the matrix $\mathbf U$).
Similarly, the orthonormality of all vectors $\mathbf v_j$ in a single calculation 
 ( by using the matrix $\mathbf V$). Then finnaly we are going to check if we obtain the result we expected and if not, we are going to explain.

In [86]:
 # u_1 is a normalised vector is if its transpose product by itself give us an identity, one everywhere on the 
# diagonal and zero otherwise.
u_1=U[:,0:1] # extraction of u_1 vector from the matrix U
np.dot(u_1.T,u_1) #

array([[1.]])

 So we can conclude that u_1 is normalised because the product give us 1 (u_1 is one imension) means identity.

In [87]:
# in the same way let's compute u_2
u_2=U[:,0:2]
u_2.T.dot(u_2)

array([[1.00000000e+00, 1.58211509e-16],
       [1.58211509e-16, 1.00000000e+00]])

We can conclude that u_2 is also normalised because $1.58211509\times 10^{-16}$ is approximatively equal to zero. So we identity and then u_2 is nomalised.

In [88]:
#Note that two vector (or matrix) are orthogonal if their product give a matrix which which the elemnts of the
# diagonal are diffrent from zero (not necessary one) and equal to zero otherwise.
u_1.T.dot(u_2) # orthogonality 

array([[1.00000000e+00, 1.58211509e-16]])

So u_1 and u_2 are orthogonal because $1.58211509\times 10^{-16}$ tend to zero.

In [89]:
# As U contain u_j, j=1,..,4, then u_j vectors are orthonormal if and ongly if U is orthonormal i.e., the product
# between the tranpose of U and U give identity.
np.dot(U.T,U)

array([[ 1.00000000e+00,  1.58211509e-16,  1.54258491e-17,
         3.98800013e-16],
       [ 1.58211509e-16,  1.00000000e+00,  4.97195777e-17,
         1.62519137e-16],
       [ 1.54258491e-17,  4.97195777e-17,  1.00000000e+00,
        -9.04317440e-17],
       [ 3.98800013e-16,  1.62519137e-16, -9.04317440e-17,
         1.00000000e+00]])

As we can observe in the matrix above, we have 1 on the main diagonal and approximatively zero everywhere, so we can easily conlude that U is orthonormal and therefore $u_j$ are orthonormal.

In [90]:
# in the same for V we have:
np.dot(V_transpose.T,V_transpose)

array([[ 1.00000000e+00,  3.62272254e-16, -1.40294248e-16,
         1.33057045e-17],
       [ 3.62272254e-16,  1.00000000e+00, -6.07900990e-17,
        -3.09912231e-16],
       [-1.40294248e-16, -6.07900990e-17,  1.00000000e+00,
         1.59691511e-16],
       [ 1.33057045e-17, -3.09912231e-16,  1.59691511e-16,
         1.00000000e+00]])

In the same we have approximatively an identity matrix so $v_j$ are orthonormal.

In short we can conclude that we don't obtain the expected results,that it's because in python we have a certain decimal after comma which is not the case theoretically. So python give us an approximate value with decimal part which is the fact that we made a certain erreur mannualy.

9.) Let's check that python has performed the SVD correctly, 
i.e., calculate the matrix $\mathbf X$  from the elements of the SVD. 
In two different ways:

(1) via the sum definition of the SVD

In [91]:
D_diag=np.diag(D)
Xsum=np.zeros([30,4])
for i in range(0,4):
    s=np.dot(U[:,i].reshape(30,1),V_transpose[i].reshape(1,4))
    Xsum+=s*D_diag[i,i]
Xsum

array([[-0.75257122, -1.67291914, -1.31966437, -1.91769058],
       [-0.2235157 , -0.59388016, -0.42257533,  0.28996009],
       [ 0.88996161, -0.23128103, -0.23685721,  0.42176013],
       [ 1.43747255,  0.45713182,  0.7307577 ,  1.21256037],
       [-1.16474238, -1.46271674, -1.00370848, -1.48934045],
       [ 0.08407471,  0.34852725,  0.78012581, -1.15984035],
       [ 0.78538087,  0.69010615,  1.30906982, -0.50084015],
       [-0.16199762,  1.01591986,  0.95173875,  1.11371034],
       [-0.39576634, -1.05107038, -0.98349107, -0.50084015],
       [-0.2173639 , -0.44849017, -0.83115519,  0.45471014],
       [ 0.63773747,  0.42384978,  0.0941442 ,  0.78421024],
       [ 2.00959073,  1.66054055,  0.92446875,  1.50911046],
       [-1.46618099, -1.78853045, -1.38501834, -1.25869038],
       [ 0.08407471, -0.45900029, -0.47006275, -0.46789014],
       [ 1.86809914,  1.14379299,  0.38094939,  0.25701008],
       [ 1.00684597,  1.51865393,  1.71435847,  0.9819103 ],
       [-0.53110612, -1.

(2) via the matrix notation of the SVD

In [92]:
U,D,V_transpose=np.linalg.svd(X_scaled,full_matrices=False)

(3) Let's verify that the obtained results are identical to the matrix $\mathbf X$:

In [93]:
np.allclose(X_scaled,Xsum) # To compare X_scaled and the results using sum definition of the SVD. 

True

In [95]:
# for matrix notation SVD
np.allclose(X_scaled,np.dot(U,np.dot(np.diag(D),V_transpose))


SyntaxError: unexpected EOF while parsing (<ipython-input-95-2820d5ba6585>, line 2)